### PR TITLE
[Core][Enable gcs scheduler 4/n] Turn on gcs scheduler feature flag

### PR DESCRIPTION
**Pull request by Chong-Li**
## Why are these changes needed?

This PR tries to turn on the gcs scheduler feature flag, and make sure it passes all test cases. Main changes include:

1. The related RayConfig (`gcs_actor_scheduling_enabled`) is set `true`.
3. In `GcsPlacementGroupScheduler::CommitBundleResources()` and `ReturnBundleResources()`, we have to trigger pending actors (in gcs) because resources have been updated.
4. Still in the above two functions, we have to update PG wildcard resources in a special way. A PG's wildcard resources (in a certain node) has to be the sum of all related bundle resources. Even though `CommitBundleResources()` uses `ToNodeBundleResourcesMap()` to sum up bundle resources, it does not handle the scenario that a single (or subset) bundle is rescheduled, in which this single bundle's wildcard resources would wrongly override the existing one.  (see `test_placement_group_reschedule_when_node_dead` for such a scenario).
5. In `GcsResourceManager::HandleGetAllResourceUsage()`, we only export gcs' pending info without adding the corresponding `ResourcesData` to the `batch` list. So as usual, the `batch` list still only contains the worker nodes.
6. Similar to raylets, GCS also periodically detects resource deadlocks, and triggers global gc if needed.
7. To pass tests like `test_actor_groups`, rpc `GetAllNodeInfo` has to return pending info in gcs additionally (do we need a dedicated rpc for that?). 
8. `test_gpu_scheduling_liveness` is temporarily skipped, because the `AffinityWithBundleSchedulingPolicy` currently does not support GPU filtering. @larrylian will submit a follow-up PR to fix this.
9. `test_autoscaler_shutdown_node_http_everynode` might not be testing the expected functionality, so it is skipped. See the `reason` for details (please correct me if there is any misunderstanding there @simon-mo ).

After the code is reviewed, I'll turn off the gcs scheduler flag to make sure this PR does not break the conventional raylet-based scheduling.

## Related issue number
#24842

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/gcs_service.proto [hunk 1]
```diff
@@ -587,7 +593,8 @@ message GetAllResourceUsageRequest {
 
 message GetAllResourceUsageReply {
   GcsStatus status = 1;
-  ResourceUsageBatchData resource_usage_data = 2;
+  bool cluster_full_of_actors_detected_by_gcs = 2;
```
Comment (DmitriGekhtman) [inline src/ray/protobuf/gcs_service.proto:596]: Could you append this as the third message rather than inserting it as the second?

Comment (DmitriGekhtman) [reply src/ray/protobuf/gcs_service.proto:596]: Same applies to all fields added in this PR.

Comment (Chong-Li) [reply src/ray/protobuf/gcs_service.proto:596]: Sure, will do.

**Comment by Chong-Li** [comment]
Is there any more suggestions to this PR? @scv119 @rkooo567

**Comment by rkooo567** [comment]
Will look into it soon. Can you remove `author-action-required` when the PR is read? (this time, I removed it! )

**Comment by rkooo567** [comment]
Do you think it is easy to separate this PR to 2~3 & others?

File: src/ray/raylet/scheduling/scheduling_ids.h [hunk 2]
```diff
@@ -185,6 +185,17 @@ class ResourceID : public BaseSchedulingID<SchedulingIDTag::Resource> {
   /// "ClusterResourceSchedulerTest" have different namespaces.
   friend void SetUnitInstanceResourceIds(absl::flat_hash_set<ResourceID> ids);
 
+  bool IsPlacementGroupWildcardResource() const {
+    auto resource_name = this->Binary();
+    std::string pattern("_group_");
```
Comment (scv119) [inline src/ray/raylet/scheduling/scheduling_ids.h:190]: @rkooo567 
is there a better way to implement this?

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -1138,6 +1138,11 @@ void NodeManager::HandleNotifyGCSRestart(const rpc::NotifyGCSRestartRequest &req
 
 void NodeManager::UpdateResourceUsage(const NodeID &node_id,
                                       const rpc::ResourcesData &resource_data) {
+  // Trigger local GC at the next heartbeat interval.
```
Comment (scv119) [inline src/ray/raylet/node_manager.cc:1141]: this seems fine to me, but why are we making this change?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 4]
```diff
@@ -189,8 +189,40 @@ void GcsServer::DoStart(const GcsInitData &gcs_init_data) {
...
+        DumpDebugStateToFile();
+
+        if (cluster_task_manager_->GetPendingQueueSize() == 0) {
+          resource_deadlock_detected_ = 0;
+          return;
+        }
+        // Trigger global gc to solve resource deadlocks.
+        // To avoid spurious triggers, only those after two consecutive
+        // detections and under throttling are sent out.
+        if (resource_deadlock_detected_++ > 0 && global_gc_throttler_->AbleToRun()) {
```
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_server.cc:205]: usually deadlock has a very precise definition. how about we rename `resource_deadlock_detected_` -> `task_pending_schedule_detected_`?

**Review by scv119 — COMMENTED** [review]
@rkooo567 to review placement group related code.

**Comment by wumuzi520** [comment]
> Do you think it is easy to separate this PR to 2~3 & others? 

+1，I think it would be more better to split it into 2~3 small ones.

**Comment by Chong-Li** [comment]
As suggested above, I'll split this PR into several small ones. Here is the plan:
1. Fix the issues related to PG scheduling (basically the item 2 & 3 in the title), also including the remaining ones from #24842 (https://github.com/ray-project/ray/pull/24842#issuecomment-1170961277).
2. With gcs scheduler enabled, the pending task info & node info have to be adapted when exported to outside modules (autoscaler, etc.). This part (item 4 & 6 in the title) will be pushed together.
3. Resolve the deadlock (long pending) issue (item 5 in the title) and all other small pieces.
Any suggestions? @rkooo567 @scv119 @wumuzi520

**Comment by rkooo567** [comment]
That split sounds perfect to me :) !!

**Pull request closed without merging**